upgrades/0.0.105.add-node-volume-encryption/upgrade.go

```diff
@@ -0,0 +1,153 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	survey "github.com/AlecAivazis/survey/v2"
+	"github.com/google/uuid"
+	"github.com/oslokommune/okctl-upgrade/upgrades/0.0.105.add-node-volume-encryption/pkg/eksctl"
+	"github.com/oslokommune/okctl-upgrade/upgrades/0.0.105.add-node-volume-encryption/pkg/lib/cmdflags"
+	"github.com/oslokommune/okctl-upgrade/upgrades/0.0.105.add-node-volume-encryption/pkg/lib/logging"
+	"github.com/oslokommune/okctl-upgrade/upgrades/0.0.105.add-node-volume-encryption/pkg/lib/manifest/apis/okctl.io/v1alpha1"
+)
+
+func upgrade(context Context, flags cmdflags.Flags, cluster v1alpha1.Cluster) error {
+	log := context.logger
+
+	if !flags.Confirm {
+		continueUpgrade, err := prompt(continuationMessage)
+		if err != nil {
+			return fmt.Errorf("prompting for continuation: %w", err)
+		}
+
+		if !continueUpgrade {
+			return errors.New("aborted by user")
+		}
+	}
+
+	log.Debug("Retrieving nodegroup names")
+
+	originalNodegroupNames, err := eksctl.GetNodeGroupNames(cluster.Metadata.Name)
+	if err != nil {
+		return fmt.Errorf("acquiring nodegroup names: %w", err)
+	}
+
+	log.Debugf("Found nodegroups %v\n", originalNodegroupNames)
+
+	clusterVersion, err := eksctl.GetClusterVersion(cluster.Metadata.Name)
+	if err != nil {
+		return fmt.Errorf("acquiring cluster version: %w", err)
+	}
+
+	log.Debugf("Found cluster version %s\n", clusterVersion)
+
+	log.Debug("Generating eksctl cluster configuration")
+
+	cfg, err := eksctl.GenerateClusterConfig(
+		cluster,
+		generateNodegroupNames(cluster.Metadata.Region, clusterVersion, generateRandomString),
+	)
+	if err != nil {
+		return fmt.Errorf("generating cluster config: %w", err)
+	}
+
+	cfg, err = configLogPrinter(log, cfg)
```

Comment:
Is it correct that the purpose of `configLogPrinter` is to log to debug the cluster config contents? Could we not simply do `fmt.Prinln(cfg.String())`, or does that move some buffer pointer, making the reader useless when attaching to the stdin of the eksctl command (in UpdateGroups)?

Regarding naming - the way I read this function name is that we create a "config log printer", for it to be used later, but what we actually do is print the config (and get a copy of it since we're using Tee). Thus perhaps it could be called "printClusterConfig"?

Reply:
The io.Reader interface doesn't support rewinding/resetting, so the stream can only be read once, hence the tee operation.

I've renamed the function